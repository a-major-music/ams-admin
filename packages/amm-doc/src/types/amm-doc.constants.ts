export const AMM_LEFT_MARGIN: number = 36;
export const AMM_LOGO_FROM_TOP: number = 120;
export const AMM_TITLE_FROM_TOP: number = 50;
export const AMM_TITLE_FROM_LEFT: number = 420;
export const AMM_ORDER_ITEM_GAPS: number[] = [40, 90, 100, 196, 30, 68];
export const AMM_ORDER_ITEM_RIGHT_ALIGN: number[] = [0, 0, 0, 1, 1, 1];
export const AMM_ORDER_ITEM_HEADER: string[] = [
  'SUPPLIER CODE',
  'UPC',
  'NAME',
  'QTY',
  'VAT',
  'TOTAL (GBP)',
];
