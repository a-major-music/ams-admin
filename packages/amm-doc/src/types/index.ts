export * from './amm-doc.type';
export * from './amm-doc.constants';
