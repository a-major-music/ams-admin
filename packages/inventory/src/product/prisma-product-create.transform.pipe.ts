import { PipeTransform, Injectable, ArgumentMetadata } from '@nestjs/common';

import * as _ from 'lodash';

/* 
The types defined in @amm/types are related, and these relationships mirror the structure of the database. However,
Prisma does not use DTOs and repositories like other ORMs. Instead it uses a JSON format to define both data and
operation in the same structure. Top level items are the same, but nested objects need operations inserted in the
object tree to tell Prisma what to do. So a create({id: 1, nested: {id: 1}}) becomes 
create({id: 1, nested: create: {id: 1}}). This interceptor inserts these operations into the objects before they hit
the controller to create variants, and to connectOrCreate suppliers and tags (to avoid duplicates)
*/
@Injectable()
export class PrismaProductCreateTransformPipe implements PipeTransform {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  transform(value: any, metadata: ArgumentMetadata) {
    const augments: any = {};

    // True if the product has more than 1 variant
    const hasVariants = value.variants.filter(v => !v.pendingDelete).length > 1;

    if (!_.isEmpty(value.suppliers)) {
      augments.suppliers = {
        connect: value.suppliers.map((s) => {
          return { guid: s.guid };
        }),
      };
    } else {
      value.suppliers = undefined;
    }

    if (!_.isEmpty(value.variants)) {
      augments.variants = { create: [] };

      // If there is only one variant we don't store filterValues and filterFields. You don't need them if
      // there is a single variant, and it improves the look of the storefront.
      value.variants.forEach((variant) => {
        if (hasVariants) {
          variant.filterValues = variant.filterValues.toString();
        }
        else {
          variant.filterValues = "";
        }

        if (!_.isEmpty(variant.imageUrls)) {
          const newField = { create: [] };
          variant.imageUrls.forEach((i) => {
            newField.create.push({ uri: i });
          });
          variant.imageUrls = newField;
        } else {
          variant.imageUrls = undefined;
        }

        augments.variants.create.push(variant);
      });
    }

    if (hasVariants) {
      augments.filterFields = value.filterFields.toString();
    }
    else {
      augments.filterFields = "";
    }

    return { ...value, ...augments };
  }
}
