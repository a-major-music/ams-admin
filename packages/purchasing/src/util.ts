export function removeProps(obj: object, props: string[]): object {
  return JSON.parse(
    JSON.stringify(obj, (k, v) => (props.includes(k) ? undefined : v))
  );
}
