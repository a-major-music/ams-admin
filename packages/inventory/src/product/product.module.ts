import { Module } from '@nestjs/common';
import { PrismaService } from '../prisma.service';
import { ProductService } from './product.service';
import { ProductController } from './product.controller';
import { HttpModule } from '@nestjs/axios';
import { ShopifyService } from '../shopify/shopify.service';

@Module({
  imports: [HttpModule],
  providers: [ProductService, ShopifyService, PrismaService],
  controllers: [ProductController],
})
export class ProductModule {}
