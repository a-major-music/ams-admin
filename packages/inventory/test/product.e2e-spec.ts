import { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { AppModule } from '../src/app.module';
import { ProductModule } from '../src/product/product.module';
import { Product, ProductType, WeightUnit } from '@amm/types';

import supertest = require('supertest');
import { PrismaService } from '../src/prisma.service';
import { removeProps } from './util';

let app: INestApplication;
let prisma: PrismaService;

beforeAll(async () => {
  const module = await Test.createTestingModule({
    imports: [AppModule, ProductModule],
  }).compile();

  app = module.createNestApplication();
  prisma = app.get<PrismaService>(PrismaService);
  await app.init();
});

afterAll(async () => {
  await app.close();
});

describe('POST /products/create', () => {
  it('should create a minimal product', async () => {
    const p: Product = {
      name: 'Testing POST single minimal product',
      taxable: true,
      productType: ProductType.INSTRUMENT,
      variants: [
        {
          sku: 'INS0001',
          weight: 1,
          weightUnit: WeightUnit.KG,
          manageStockLevels: true,
          stockOnHand: 0,
          movingAverageCost: 500,
          sellable: true,
          buyable: true,
          retailPrice: 1000,
          buyPrice: 500,
          imageUrls: [
            'https://cdn.shopify.com/s/files/1/1079/2732/products/musam969056_0.jpg?v=1647272012',
          ],
        },
      ],
    };

    const cs = await supertest
      .agent(app.getHttpServer())
      .post('/products/create')
      .set('Accept', 'application/json')
      .send(p)
      .expect('Content-Type', /json/)
      .expect(201);

    const qs = await supertest
      .agent(app.getHttpServer())
      .get(`/products/${cs.body.guid}`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(JSON.parse(qs.text), ['guid'])).toMatchSnapshot();

    await prisma.productVariant.deleteMany({});
    await prisma.product.deleteMany({});
  });

  it('should create a single, fully populated product', async () => {
    const p: Product = {
      name: 'Testing POST single product',
      taxable: true,
      productType: ProductType.INSTRUMENT,
      brand: 'Test Brand',
      description: 'Test Description',
      tags: ['tag1'],
      // variants should be created
      variants: [
        {
          filterValues: 'large',
          barcode: '1234567890123',
          supplierCode: 'SUP101',
          buyPrice: 1000,
          buyable: true,
          manageStockLevels: true,
          notes: 'A test variant',
          previousPrice: 2000,
          retailPrice: 1500,
          sellable: true,
          sku: 'INS00001',
          weight: 5,
          weightUnit: WeightUnit.KG,
          stockOnHand: 0,
          movingAverageCost: 500,
        },
      ],
      filterFields: 'size',
      location: 'A1',
      supplierGuid: 'supplier-fake-guid',
    };

    const cs = await supertest
      .agent(app.getHttpServer())
      .post('/products/create')
      .set('Accept', 'application/json')
      .send(p)
      .expect('Content-Type', /json/)
      .expect(201);

    const qs = await supertest
      .agent(app.getHttpServer())
      .get(`/products/${cs.body.guid}`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(JSON.parse(qs.text), ['guid'])).toMatchSnapshot();

    const countTest = await supertest
      .agent(app.getHttpServer())
      .get(`/products/count`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(countTest.body).toMatchSnapshot();

    await prisma.productVariant.deleteMany({});
    await prisma.product.deleteMany({});
  });
});

describe('PUT /products/:id/update', () => {
  // Create dummy product
  const p: Product = {
    name: 'Testing add new variant to product',
    productType: ProductType.INSTRUMENT,
    brand: 'Test Brand',
    description: 'Test Description',
    filterFields: 'size',
    location: 'A1',
  };

  it('should add a new variant to a product', async () => {
    const pRes = await supertest
      .agent(app.getHttpServer())
      .post('/products/create')
      .send(p)
      .expect('Content-Type', /json/)
      .expect(201);

    const pv = {
      variants: [
        {
          filterValues: 'large',
          barcode: '1234567890123',
          supplierCode: 'SUP101',
          buyPrice: 1000,
          buyable: true,
          manageStockLevels: true,
          notes: 'A test variant',
          previousPrice: 2000,
          retailPrice: 1500,
          sellable: true,
          sku: 'INS00001',
          weight: 5,
          weightUnit: WeightUnit.KG,
          stockOnHand: 0,
          movingAverageCost: 500,
        },
      ],
    };

    p.guid = pRes.body.guid;

    await supertest
      .agent(app.getHttpServer())
      .put(`/products/${p.guid}/update`)
      .send(pv)
      .expect('Content-Type', /json/)
      .expect(200);

    const checkRes = await supertest
      .agent(app.getHttpServer())
      .get(`/products/${pRes.body.guid}`)
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(checkRes.body, ['guid'])).toMatchSnapshot();

    // Store in p for later tests in this group
    p.variants = checkRes.body.variants;
  });

  it('should add a tag', async () => {
    await supertest
      .agent(app.getHttpServer())
      .put(`/products/${p.guid}/update`)
      .send({ tags: ['tag1'] })
      .expect('Content-Type', /json/)
      .expect(200);

    const checkRes = await supertest
      .agent(app.getHttpServer())
      .get(`/products/${p.guid}`)
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(checkRes.body, ['guid'])).toMatchSnapshot();

    p.tags = checkRes.body.tags;
  });

  it('should add another tag', async () => {
    await supertest
      .agent(app.getHttpServer())
      .put(`/products/${p.guid}/update`)
      .send({ tags: ['tag1', 'tag2'] })
      .expect('Content-Type', /json/)
      .expect(200);

    const checkRes = await supertest
      .agent(app.getHttpServer())
      .get(`/products/${p.guid}`)
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(checkRes.body, ['guid'])).toMatchSnapshot();

    p.tags = checkRes.body.tags;
  });

  it('should update a tag', async () => {
    await supertest
      .agent(app.getHttpServer())
      .put(`/products/${p.guid}/update`)
      .send({ tags: ['tag1', 'tag3'] })
      .expect('Content-Type', /json/)
      .expect(200);

    const checkRes = await supertest
      .agent(app.getHttpServer())
      .get(`/products/${p.guid}`)
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(checkRes.body, ['guid'])).toMatchSnapshot();

    p.tags = checkRes.body.tags;
  });

  it('should delete all tags', async () => {
    await supertest
      .agent(app.getHttpServer())
      .put(`/products/${p.guid}/update`)
      .send({ tags: [] })
      .expect('Content-Type', /json/)
      .expect(200);

    const checkRes = await supertest
      .agent(app.getHttpServer())
      .get(`/products/${p.guid}`)
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(checkRes.body, ['guid'])).toMatchSnapshot();

    p.tags = checkRes.body.tags;
  });

  it('should delete a variant', async () => {
    const pRes = await supertest
      .agent(app.getHttpServer())
      .put(`/products/${p.guid}/update`)
      .send({ variants: [{ ...p.variants[0], pendingDelete: true }] })
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(pRes.body, ['guid'])).toMatchSnapshot();

    await prisma.productVariant.deleteMany({});
    await prisma.product.deleteMany({});
  });
});

describe('GET /products/:guid', () => {
  it('should return a single product', async () => {
    const p = {
      name: 'Testing GET single product',
      taxable: true,
      productType: ProductType.INSTRUMENT,
      brand: 'Test Brand',
      description: 'Test Description',
      tags: ['instrument', 'test'],
      variants: [
        {
          filterValues: 'large',
          barcode: '1234567890123',
          supplierCode: 'SUP101',
          buyPrice: 1000,
          buyable: true,
          manageStockLevels: true,
          notes: 'A test variant',
          previousPrice: 2000,
          retailPrice: 1500,
          sellable: true,
          sku: 'INS00001',
          weight: 5,
          weightUnit: WeightUnit.KG,
          stockOnHand: 0,
          movingAverageCost: 500,
        },
      ],
      filterFields: 'size',
      location: 'A1',
    };

    const cs = await supertest
      .agent(app.getHttpServer())
      .post('/products/create')
      .set('Accept', 'application/json')
      .send(p)
      .expect('Content-Type', /json/)
      .expect(201);

    const { body } = await supertest
      .agent(app.getHttpServer())
      .get(`/products/${cs.body.guid}`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(body, ['guid'])).toMatchSnapshot();

    await prisma.productVariant.deleteMany({});
    await prisma.product.deleteMany({});
  });
});

describe('GET /products', () => {
  it('should return an array of products', async () => {
    await prisma.product.create({
      data: {
        name: 'Testing GET products (product 1)',
        filterFields: 'size',
        location: 'A1',
        productType: ProductType.BOOK,
        brand: 'Test Brand',
        description: 'Test description - product 1',
      },
    });

    await prisma.product.create({
      data: {
        name: 'Testing GET products (product 2)',
        filterFields: 'size',
        location: 'A1',
        productType: ProductType.BOOK,
        brand: 'Test Brand',
        description: 'Test description - product 2',
      },
    });

    // Test /products
    const { body } = await supertest
      .agent(app.getHttpServer())
      .get('/products')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(body, ['guid'])).toMatchSnapshot();

    await prisma.product.deleteMany({});
  });
});

describe('GET /products/variants/:sku', () => {
  const p = {
    name: 'Testing GET products by variant sku',
    filterFields: 'size',
    location: 'A1',
    productType: ProductType.BOOK,
    brand: 'Test Brand',
    description: 'Test description',
    variants: [
      {
        filterValues: 'large',
        barcode: '1234567890123',
        supplierCode: 'SUP101',
        buyPrice: 1000,
        buyable: true,
        manageStockLevels: true,
        notes: 'A test variant',
        previousPrice: 2000,
        retailPrice: 1500,
        sellable: true,
        sku: 'INS00001',
        weight: 5,
        weightUnit: WeightUnit.KG,
        stockOnHand: 0,
        movingAverageCost: 500,
      },
    ],
  };

  it('should return no products', async () => {
    await supertest
      .agent(app.getHttpServer())
      .post('/products/create')
      .set('Accept', 'application/json')
      .send(p)
      .expect('Content-Type', /json/)
      .expect(201);

    const { body } = await supertest
      .agent(app.getHttpServer())
      .get('/products/variant/NO_SUCH_SKU')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(body, ['guid'])).toMatchSnapshot();
  });

  it('should return a variant', async () => {
    const { body } = await supertest
      .agent(app.getHttpServer())
      .get(`/products/variant/${p.variants[0].sku}/?variantOnly=true`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(body, ['guid'])).toMatchSnapshot();
  });

  it('should return a variant map', async () => {
    const { body } = await supertest
      .agent(app.getHttpServer())
      .get(`/products/variants-guid-map`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(body[0]).toBeDefined();
    expect(body[0].guid).toBeDefined();
    expect(body[0].sku).toEqual(p.variants[0].sku);
  });

  it('should return a product', async () => {
    const { body } = await supertest
      .agent(app.getHttpServer())
      .get(`/products/variant/${p.variants[0].sku}`)
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);

    expect(removeProps(body, ['guid'])).toMatchSnapshot();

    await prisma.productVariant.deleteMany({});
    await prisma.product.deleteMany({});
  });

  describe.skip('GET /products?search', () => {
    const p = {
      name: 'Testing GET products by variant sku',
      filterFields: 'size',
      location: 'A1',
      productType: ProductType.BOOK,
      brand: 'Test Brand',
      description: 'Test description',
      variants: [
        {
          filterValues: 'large',
          barcode: '1234567890123',
          supplierCode: 'SUP101',
          buyPrice: 1000,
          buyable: true,
          manageStockLevels: true,
          notes: 'A test variant',
          previousPrice: 2000,
          retailPrice: 1500,
          sellable: true,
          sku: 'INS00001',
          weight: 5,
          weightUnit: WeightUnit.KG,
          stockOnHand: 0,
          movingAverageCost: 500,
        },
      ],
    };

    it('should return products by partial name, partial sku or barcode', async () => {
      await supertest
        .agent(app.getHttpServer())
        .post('/products/create')
        .set('Accept', 'application/json')
        .send(p)
        .expect('Content-Type', /json/)
        .expect(201);

      // Should return product by partial SKU
      const r1 = await supertest
        .agent(app.getHttpServer())
        .get('/products?search=INS')
        .set('Accept', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200);

      expect(removeProps(r1.body, ['guid'])).toMatchSnapshot();

      // Should return product by partial name
      const r2 = await supertest
        .agent(app.getHttpServer())
        .get('/products?search=Test')
        .set('Accept', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200);

      expect(removeProps(r2.body, ['guid'])).toMatchSnapshot();

      // Should return nothing
      const r3 = await supertest
        .agent(app.getHttpServer())
        .get('/products?search=Wibble')
        .set('Accept', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200);

      expect(removeProps(r3.body, ['guid'])).toMatchSnapshot();

      // Should return product by partial barcode
      const r4 = await supertest
        .agent(app.getHttpServer())
        .get('/products?search=12345')
        .set('Accept', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200);

      expect(removeProps(r4.body, ['guid'])).toMatchSnapshot();

      await prisma.productVariant.deleteMany({});
      await prisma.product.deleteMany({});
    });
  });
});
