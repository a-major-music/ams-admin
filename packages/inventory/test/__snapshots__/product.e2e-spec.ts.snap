// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /products should return an array of products 1`] = `
Array [
  Object {
    "brand": "Test Brand",
    "description": "Test description - product 1",
    "filterFields": "size",
    "id": 5,
    "location": "A1",
    "name": "Testing GET products (product 1)",
    "productType": "Book",
    "publishedToShopify": false,
    "shopifyId": null,
    "supplierGuid": null,
    "tags": Array [],
    "taxable": false,
    "variants": Array [],
  },
  Object {
    "brand": "Test Brand",
    "description": "Test description - product 2",
    "filterFields": "size",
    "id": 6,
    "location": "A1",
    "name": "Testing GET products (product 2)",
    "productType": "Book",
    "publishedToShopify": false,
    "shopifyId": null,
    "supplierGuid": null,
    "tags": Array [],
    "taxable": false,
    "variants": Array [],
  },
]
`;

exports[`GET /products/:guid should return a single product 1`] = `
Object {
  "brand": "Test Brand",
  "description": "Test Description",
  "filterFields": "size",
  "id": 4,
  "location": "A1",
  "name": "Testing GET single product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": null,
  "supplierGuid": null,
  "tags": Array [
    "instrument",
    "test",
  ],
  "taxable": true,
  "variants": Array [
    Object {
      "barcode": "1234567890123",
      "boughtInPacks": false,
      "buyPrice": 1000,
      "buyable": true,
      "filterValues": "large",
      "id": 4,
      "imageUrls": Array [],
      "manageStockLevels": true,
      "movingAverageCost": 500,
      "notes": "A test variant",
      "packSize": null,
      "previousPrice": 2000,
      "productId": 4,
      "retailPrice": 1500,
      "sellable": true,
      "sku": "INS00001",
      "stockOnHand": 0,
      "supplierCode": "SUP101",
      "weight": 5,
      "weightUnit": "Kg",
    },
  ],
}
`;

exports[`GET /products/variants/:sku GET /products?search should return products by partial name, partial sku or barcode 1`] = `
Array [
  Object {
    "brand": "Test Brand",
    "description": "Test description",
    "filterFields": "size",
    "id": 10,
    "location": "A1",
    "name": "Testing GET products by variant sku",
    "productType": "Book",
    "suppliers": Array [],
    "tags": Array [],
    "taxable": false,
    "variants": Array [
      Object {
        "barcode": "1234567890123",
        "boughtInPacks": false,
        "buyPrice": 1000,
        "buyable": true,
        "filterValues": "large",
        "id": 8,
        "imageUrls": Array [],
        "manageStockLevels": true,
        "movingAverageCost": 500,
        "notes": "A test variant",
        "packSize": null,
        "previousPrice": 2000,
        "productId": 10,
        "retailPrice": 1500,
        "sellable": true,
        "sku": "INS00001",
        "stockOnHand": 0,
        "supplierCode": "SUP101",
        "weight": 5,
        "weightUnit": "Kg",
      },
    ],
  },
]
`;

exports[`GET /products/variants/:sku GET /products?search should return products by partial name, partial sku or barcode 2`] = `
Array [
  Object {
    "brand": "Test Brand",
    "description": "Test description",
    "filterFields": "size",
    "id": 10,
    "location": "A1",
    "name": "Testing GET products by variant sku",
    "productType": "Book",
    "suppliers": Array [],
    "tags": Array [],
    "taxable": false,
    "variants": Array [
      Object {
        "barcode": "1234567890123",
        "boughtInPacks": false,
        "buyPrice": 1000,
        "buyable": true,
        "filterValues": "large",
        "id": 8,
        "imageUrls": Array [],
        "manageStockLevels": true,
        "movingAverageCost": 500,
        "notes": "A test variant",
        "packSize": null,
        "previousPrice": 2000,
        "productId": 10,
        "retailPrice": 1500,
        "sellable": true,
        "sku": "INS00001",
        "stockOnHand": 0,
        "supplierCode": "SUP101",
        "weight": 5,
        "weightUnit": "Kg",
      },
    ],
  },
]
`;

exports[`GET /products/variants/:sku GET /products?search should return products by partial name, partial sku or barcode 3`] = `Array []`;

exports[`GET /products/variants/:sku GET /products?search should return products by partial name, partial sku or barcode 4`] = `
Array [
  Object {
    "brand": "Test Brand",
    "description": "Test description",
    "filterFields": "size",
    "id": 10,
    "location": "A1",
    "name": "Testing GET products by variant sku",
    "productType": "Book",
    "suppliers": Array [],
    "tags": Array [],
    "taxable": false,
    "variants": Array [
      Object {
        "barcode": "1234567890123",
        "boughtInPacks": false,
        "buyPrice": 1000,
        "buyable": true,
        "filterValues": "large",
        "id": 8,
        "imageUrls": Array [],
        "manageStockLevels": true,
        "movingAverageCost": 500,
        "notes": "A test variant",
        "packSize": null,
        "previousPrice": 2000,
        "productId": 10,
        "retailPrice": 1500,
        "sellable": true,
        "sku": "INS00001",
        "stockOnHand": 0,
        "supplierCode": "SUP101",
        "weight": 5,
        "weightUnit": "Kg",
      },
    ],
  },
]
`;

exports[`GET /products/variants/:sku should return a product 1`] = `
Array [
  Object {
    "brand": "Test Brand",
    "description": "Test description",
    "filterFields": "size",
    "id": 7,
    "location": "A1",
    "name": "Testing GET products by variant sku",
    "productType": "Book",
    "publishedToShopify": false,
    "shopifyId": null,
    "supplierGuid": null,
    "tags": Array [],
    "taxable": false,
    "variants": Array [
      Object {
        "barcode": "1234567890123",
        "boughtInPacks": false,
        "buyPrice": 1000,
        "buyable": true,
        "filterValues": "large",
        "id": 5,
        "imageUrls": Array [],
        "manageStockLevels": true,
        "movingAverageCost": 500,
        "notes": "A test variant",
        "packSize": null,
        "previousPrice": 2000,
        "productId": 7,
        "retailPrice": 1500,
        "sellable": true,
        "sku": "INS00001",
        "stockOnHand": 0,
        "supplierCode": "SUP101",
        "weight": 5,
        "weightUnit": "Kg",
      },
    ],
  },
]
`;

exports[`GET /products/variants/:sku should return a variant 1`] = `
Object {
  "barcode": "1234567890123",
  "boughtInPacks": false,
  "buyPrice": 1000,
  "buyable": true,
  "filterValues": "large",
  "id": 5,
  "manageStockLevels": true,
  "movingAverageCost": 500,
  "notes": "A test variant",
  "packSize": null,
  "previousPrice": 2000,
  "productId": 7,
  "retailPrice": 1500,
  "sellable": true,
  "sku": "INS00001",
  "stockOnHand": 0,
  "supplierCode": "SUP101",
  "weight": 5,
  "weightUnit": "Kg",
}
`;

exports[`GET /products/variants/:sku should return no products 1`] = `Array []`;

exports[`POST /products/create should create a minimal product 1`] = `
Object {
  "brand": null,
  "description": null,
  "filterFields": null,
  "id": 1,
  "location": "Unknown",
  "name": "Testing POST single minimal product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": null,
  "supplierGuid": null,
  "tags": Array [],
  "taxable": true,
  "variants": Array [
    Object {
      "barcode": null,
      "boughtInPacks": false,
      "buyPrice": 500,
      "buyable": true,
      "filterValues": null,
      "id": 1,
      "imageUrls": Array [
        Object {
          "id": 1,
          "productVariantId": 1,
          "uri": "https://cdn.shopify.com/s/files/1/1079/2732/products/musam969056_0.jpg?v=1647272012",
        },
      ],
      "manageStockLevels": true,
      "movingAverageCost": 500,
      "notes": null,
      "packSize": null,
      "previousPrice": null,
      "productId": 1,
      "retailPrice": 1000,
      "sellable": true,
      "sku": "INS0001",
      "stockOnHand": 0,
      "supplierCode": null,
      "weight": 1,
      "weightUnit": "Kg",
    },
  ],
}
`;

exports[`POST /products/create should create a single, fully populated product 1`] = `
Object {
  "brand": "Test Brand",
  "description": "Test Description",
  "filterFields": "size",
  "id": 2,
  "location": "A1",
  "name": "Testing POST single product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": null,
  "supplierGuid": "supplier-fake-guid",
  "tags": Array [
    "tag1",
  ],
  "taxable": true,
  "variants": Array [
    Object {
      "barcode": "1234567890123",
      "boughtInPacks": false,
      "buyPrice": 1000,
      "buyable": true,
      "filterValues": "large",
      "id": 2,
      "imageUrls": Array [],
      "manageStockLevels": true,
      "movingAverageCost": 500,
      "notes": "A test variant",
      "packSize": null,
      "previousPrice": 2000,
      "productId": 2,
      "retailPrice": 1500,
      "sellable": true,
      "sku": "INS00001",
      "stockOnHand": 0,
      "supplierCode": "SUP101",
      "weight": 5,
      "weightUnit": "Kg",
    },
  ],
}
`;

exports[`POST /products/create should create a single, fully populated product 2`] = `
Object {
  "rows": 1,
}
`;

exports[`PUT /products/:id/update should add a new variant to a product 1`] = `
Object {
  "brand": "Test Brand",
  "description": "Test Description",
  "filterFields": "size",
  "id": 3,
  "location": "A1",
  "name": "Testing add new variant to product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": "",
  "supplierGuid": null,
  "tags": Array [],
  "taxable": false,
  "variants": Array [
    Object {
      "barcode": "1234567890123",
      "boughtInPacks": false,
      "buyPrice": 1000,
      "buyable": true,
      "filterValues": "large",
      "id": 3,
      "imageUrls": Array [],
      "manageStockLevels": true,
      "movingAverageCost": 500,
      "notes": "A test variant",
      "packSize": null,
      "previousPrice": 2000,
      "productId": 3,
      "retailPrice": 1500,
      "sellable": true,
      "sku": "INS00001",
      "stockOnHand": 0,
      "supplierCode": "SUP101",
      "weight": 5,
      "weightUnit": "Kg",
    },
  ],
}
`;

exports[`PUT /products/:id/update should add a tag 1`] = `
Object {
  "brand": "Test Brand",
  "description": "Test Description",
  "filterFields": "size",
  "id": 3,
  "location": "A1",
  "name": "Testing add new variant to product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": "",
  "supplierGuid": null,
  "tags": Array [
    "tag1",
  ],
  "taxable": false,
  "variants": Array [
    Object {
      "barcode": "1234567890123",
      "boughtInPacks": false,
      "buyPrice": 1000,
      "buyable": true,
      "filterValues": "large",
      "id": 3,
      "imageUrls": Array [],
      "manageStockLevels": true,
      "movingAverageCost": 500,
      "notes": "A test variant",
      "packSize": null,
      "previousPrice": 2000,
      "productId": 3,
      "retailPrice": 1500,
      "sellable": true,
      "sku": "INS00001",
      "stockOnHand": 0,
      "supplierCode": "SUP101",
      "weight": 5,
      "weightUnit": "Kg",
    },
  ],
}
`;

exports[`PUT /products/:id/update should add another tag 1`] = `
Object {
  "brand": "Test Brand",
  "description": "Test Description",
  "filterFields": "size",
  "id": 3,
  "location": "A1",
  "name": "Testing add new variant to product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": "",
  "supplierGuid": null,
  "tags": Array [
    "tag1",
    "tag2",
  ],
  "taxable": false,
  "variants": Array [
    Object {
      "barcode": "1234567890123",
      "boughtInPacks": false,
      "buyPrice": 1000,
      "buyable": true,
      "filterValues": "large",
      "id": 3,
      "imageUrls": Array [],
      "manageStockLevels": true,
      "movingAverageCost": 500,
      "notes": "A test variant",
      "packSize": null,
      "previousPrice": 2000,
      "productId": 3,
      "retailPrice": 1500,
      "sellable": true,
      "sku": "INS00001",
      "stockOnHand": 0,
      "supplierCode": "SUP101",
      "weight": 5,
      "weightUnit": "Kg",
    },
  ],
}
`;

exports[`PUT /products/:id/update should delete a variant 1`] = `
Object {
  "brand": "Test Brand",
  "description": "Test Description",
  "filterFields": "size",
  "id": 3,
  "location": "A1",
  "name": "Testing add new variant to product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": "",
  "supplierGuid": null,
  "tags": Array [],
  "taxable": false,
  "variants": Array [],
}
`;

exports[`PUT /products/:id/update should delete all tags 1`] = `
Object {
  "brand": "Test Brand",
  "description": "Test Description",
  "filterFields": "size",
  "id": 3,
  "location": "A1",
  "name": "Testing add new variant to product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": "",
  "supplierGuid": null,
  "tags": Array [],
  "taxable": false,
  "variants": Array [
    Object {
      "barcode": "1234567890123",
      "boughtInPacks": false,
      "buyPrice": 1000,
      "buyable": true,
      "filterValues": "large",
      "id": 3,
      "imageUrls": Array [],
      "manageStockLevels": true,
      "movingAverageCost": 500,
      "notes": "A test variant",
      "packSize": null,
      "previousPrice": 2000,
      "productId": 3,
      "retailPrice": 1500,
      "sellable": true,
      "sku": "INS00001",
      "stockOnHand": 0,
      "supplierCode": "SUP101",
      "weight": 5,
      "weightUnit": "Kg",
    },
  ],
}
`;

exports[`PUT /products/:id/update should update a tag 1`] = `
Object {
  "brand": "Test Brand",
  "description": "Test Description",
  "filterFields": "size",
  "id": 3,
  "location": "A1",
  "name": "Testing add new variant to product",
  "productType": "Instrument",
  "publishedToShopify": false,
  "shopifyId": "",
  "supplierGuid": null,
  "tags": Array [
    "tag1",
    "tag3",
  ],
  "taxable": false,
  "variants": Array [
    Object {
      "barcode": "1234567890123",
      "boughtInPacks": false,
      "buyPrice": 1000,
      "buyable": true,
      "filterValues": "large",
      "id": 3,
      "imageUrls": Array [],
      "manageStockLevels": true,
      "movingAverageCost": 500,
      "notes": "A test variant",
      "packSize": null,
      "previousPrice": 2000,
      "productId": 3,
      "retailPrice": 1500,
      "sellable": true,
      "sku": "INS00001",
      "stockOnHand": 0,
      "supplierCode": "SUP101",
      "weight": 5,
      "weightUnit": "Kg",
    },
  ],
}
`;
